[master] Matchers can self-reference the loader

frebib opened the pull request:
This makes it so a loaded matcher doesn't have to load another instance of the loader itself and can instead reuse the existing matchers that are already loaded. This should speed up many matcher operations considerably.

This is an alternative approach to https://github.com/saltstack/salt/pull/64606 which performs the same but is much much cleaner

### What does this PR do?

### What issues does this PR fix or reference?
Fixes: #61950

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

frebib commented:
Okay yeah this works and it drops our pillar topfile matching from ~4-5s to <200ms

s0undt3ch commented:
Can you please address the `pre-commit` issues(https://docs.saltproject.io/en/latest/topics/development/contributing.html#set-up-pre-commit-and-nox), and, for lint, please also add `__matchers__` to the [`.pylintrc` file](https://github.com/saltstack/salt/blob/v3006.1/.pylintrc#L355) under `additional-builtins`.

Thanks!

frebib commented:
@s0undt3ch This also requires a change to pytest-salt-factories because that hardcodes a list of dunders. We might also need to add this new dunder to a few of the other loader functions too, like `minion_mods()`. Starting with pytest-salt-factories, should I send a PR there first? It'll be required to make the tests pass

frebib commented:
fyi half of the pre-commit hooks fail on python3.11 because you're pinning ancient library versions that don't support 3.11, mostly because they're trying to import `formatargspec` and other deprecated functions
```
  File "/home/frebib/.cache/pre-commit/repoe0j2g1y8/py_env-python3/lib/python3.11/site-packages/wrapt/__init__.py", line 10, in <module>
    from .decorators import (adapter_factory, AdapterFactory, decorator,
  File "/home/frebib/.cache/pre-commit/repoe0j2g1y8/py_env-python3/lib/python3.11/site-packages/wrapt/decorators.py", line 34, in <module>
    from inspect import ismethod, isclass, formatargspec
ImportError: cannot import name 'formatargspec' from 'inspect' (/usr/lib/python3.11/inspect.py)
nox > Command pylint --rcfile=.pylintrc --disable=I tests/support/pytest/loader.py failed with exit code 1
```

frebib committed:
Matchers can self-reference the loader

This makes it so a loaded matcher doesn't have to load another instance
of the loader itself and can instead reuse the existing matchers that
are already loaded. This should speed up many matcher operations
considerably.

Signed-off-by: Joe Groocock <jgroocock@cloudflare.com>

frebib commented on salt/loader/__init__.py:
```diff
@@ -146,7 +146,7 @@ def _module_dirs(
     if ext_dirs:
         if ext_type_dirs is None:
-            ext_type_dirs = "{}_dirs".format(tag)
+            ext_type_dirs = f"{tag}_dirs"
```
Annoyingly pre-commit has made a mess of this patch including unrelated changes.
It would be better to pyupgrade/black/isort the whole repo in a single commit so contributor patches can be as clean as possible

dwoz replied:
I agree this could be done in a bulk or several smaller PRs. That said, for now you could also separate the your changes from pre-commit's in two separate commits. I think `--no-verify` can accomplish that.

dwoz reviewed (changes requested):
I'm good with this when it has more tests. :+1:

twangboy reviewed (changes requested):
This needs a few tests. And a rebase.

twangboy reviewed (changes requested):
Please create a changelog for this

dwoz committed:
Add changelog and expand tests for matchers self-reference loader

Add changelog entry for PR #64607 and replace the now-obsolete
test_matchers_from_context test (which tested __context__ caching) with
tests that verify __matchers__ is injected and never causes recursive
salt.loader.matchers() calls.

dwoz committed:
Fix black formatting in compound_match.py
